core/src/ledger/vp_env.rs

```diff
@@ -97,6 +106,45 @@ where
     /// Get a tx hash
     fn get_tx_code_hash(&self) -> Result<Option<Hash>, storage_api::Error>;
 
+    /// Get the shielded action including the transfer and the masp tx
+    fn get_shielded_action(
+        &self,
+        tx_data: Tx,
+    ) -> Result<(Transfer, Transaction), storage_api::Error> {
+        let signed = tx_data;
+        match Transfer::try_from_slice(&signed.data().unwrap()[..]) {
+            Ok(transfer) => {
+                let shielded_hash = transfer
+                    .shielded
+                    .ok_or_err_msg("unable to find shielded hash")?;
+                let masp_tx = signed
+                    .get_section(&shielded_hash)
+                    .and_then(|x| x.as_ref().masp_tx())
+                    .ok_or_err_msg("unable to find shielded section")?;
+                Ok((transfer, masp_tx))
+            }
+            Err(_) => {
+                if let Some(event) =
+                    self.get_ibc_event("fungible_token_packet".to_string())?
+                {
+                    if let Some(shielded) =
+                        get_shielded_transfer(&event).into_storage_result()?
+                    {
+                        Ok((shielded.transfer, shielded.masp_tx))
+                    } else {
+                        Err(storage_api::Error::new_const(
+                            "No shielded transfer in the IBC event",
+                        ))
+                    }
+                } else {
+                    Err(storage_api::Error::new_const(
+                        "No IBC event for the shielded action",
+                    ))
+                }
+            }
+        }
```

Comment:
I think this block of code could be improved. IMO it doesn't read very well. if possible, I'd like it to remain somewhat flat (at at most two indentation levels)